terraso_backend/apps/collaboration/models/memberships.py

```diff
@@ -0,0 +1,167 @@
+
+        membership = self.get_membership(user)
+        is_new = not membership
+
+        previous_membership_status = membership.membership_status if not is_new else None
+
+        if validation_func(
+            {
+                "user_role": user_role,
+                "membership_status": membership_status,
+                "current_membership": membership,
+            }
+        ):
+            raise ValidationError("User cannot request membership")
+
+        if is_new:
+            membership = Membership(
+                membership_list=self,
+                user=user,
+                user_role=user_role,
+                membership_status=membership_status,
+            )
+        else:
+            membership.user_role = user_role
+            if membership_status:
+                membership.membership_status = membership_status
+
+        membership_status = membership.membership_status
+
+        is_membership_approved = (
+            previous_membership_status is not None
+            and previous_membership_status != Membership.APPROVED
+            and membership_status == Membership.APPROVED
+        )
+
+        membership.save()
+        return is_membership_approved, membership
+
+    @property
+    def approved_members(self):
+        approved_memberships_user_ids = models.Subquery(
+            self.memberships.approved_only().values("user_id")
+        )
+        return User.objects.filter(id__in=approved_memberships_user_ids)
+
+    def has_role(self, user: User, role: str) -> bool:
+        return self.memberships.by_role(role).filter(id=user.id).exists()
+
+    def is_member(self, user: User) -> bool:
+        return self.approved_members.filter(id=user.id).exists()
+
+    def get_membership(self, user: User):
+        return self.memberships.filter(user=user).first()
+
+    @property
+    def can_join(self):
+        return self.enroll_method in (self.ENROLL_METHOD_JOIN, self.ENROLL_METHOD_BOTH)
+
+    @classmethod
+    def get_membership_type_from_text(cls, membership_type):
+        if membership_type and membership_type.lower() == cls.MEMBERSHIP_TYPE_CLOSED:
+            return cls.MEMBERSHIP_TYPE_CLOSED
+        return cls.MEMBERSHIP_TYPE_OPEN
+
+
+class MembershipObjectsManager(SafeDeleteManager):
+    def by_role(self, role):
+        return self.filter(user_role=role, membership_status=Membership.APPROVED)
+
+    def approved_only(self):
+        return self.filter(membership_status=Membership.APPROVED)
+
+
+class Membership(BaseModel):
+    APPROVED = "approved"
+    PENDING = "pending"
+
+    APPROVAL_STATUS = (
+        (APPROVED, _("Approved")),
+        (PENDING, _("Pending")),
+    )
+
+    membership_list = models.ForeignKey(
+        MembershipList, on_delete=models.CASCADE, related_name="memberships"
+    )
+    user = models.ForeignKey(
+        User, on_delete=models.CASCADE, related_name="collaboration_memberships"
+    )
+
+    user_role = models.CharField(max_length=64)
```

Comment:
issue: there is no validation that the role field is valid for the given membership with the current implementation, so a typo in a front end query could lead to corrupt data in the DB. this is the biggest thing that bugs me about the implementation because if we do end up wanting/needing to change it in the future (which seems very possible) it will require a data migration, all the other things will be much easier to change in code

the two ways i see to implement this as a database constraint would be:
- don't have a role field on `Membership`. models can define their own membership role/permission types by creating 1-1 fields pointing at membership. other models might not even want to have the shape of their permissions fields be a char choice field (e.g. it might be several boolean permission fields)
- turn this into a `MembershipListBase` and `MembershipBase` class and having each model that needs memberships extend them

or, without a database constraint:
- the `save_membership` function in `MembershipList` could accept an additional required parameter `valid_roles`
- continue as is, where each model using `MembershipList` must remember to do their own validation which will not be guaranteed by a DB constraint. this feels brittle to me hence the comment

Reply:
ok that sounds good, i agree DB constraints would be the most work to implement and i'm ok with just having application layer validation right now

also thinking about it more, i don't see a future situation that would _require_ a data migration if we go ahead with this implementation, since if we either wanted a DB constraint or permissions-based role model for a particular user of MembershipList, they could still implement their data extension in the way i described in the first bullet point and just ignore `Membership`'s role field. there would be a redundant column which wouldn't be elegant but not that big a deal

so i don't think this implementation would ever put us in a situation where we need to do a big refactor/data migration which is good & i feel good about the direction you're headed!